spec/models/tag_spec.rb

```diff
@@ -0,0 +1,117 @@
+require "rails_helper"
+
+describe Tag, type: :model do
+  let(:tag) {FactoryBot.create :tag}
+
+  describe "factories" do
+    subject {tag}
+    it {is_expected.to be_valid}
+  end
+
+  describe "associations" do
+    it {is_expected.to have_many :tag_labels}
+    it {is_expected.to have_many(:labels).through :tag_labels}
+    it {is_expected.to have_many :tag_users}
+    it {is_expected.to have_many(:users).through :tag_users}
+    it {is_expected.to have_many :comments}
+    it {is_expected.to have_many :attachments}
+    it {is_expected.to have_many :checklists}
+    it {is_expected.to belong_to :list}
+  end
+
+  describe "validations" do
+    context "when missing name" do
+      before {tag.name = nil}
+      it {is_expected.to validate_presence_of :name}
+    end
+
+    context "when name length is greater than #{Settings.tag.name.length} characters" do
+      before {tag.name = Faker::Name.initials number: Settings.tag.name.length + 1}
+      it {is_expected.to validate_length_of(:name).is_at_most Settings.tag.name.length}
+    end
+
+    context "when description length is greater than #{Settings.tag.description.length} characters" do
+      before {tag.description = Faker::Name.initials number: Settings.tag.description.length + 1}
+      it {is_expected.to validate_length_of(:description).is_at_most Settings.tag.description.length}
+    end
+
+    context "when deadline format is invalid" do
+      before {tag.deadline = "1f/5t/2020"}
+      it {is_expected.to_not allow_value("1f/5t/2020").for :deadline}
+    end
+  end
+
+  describe "scopes" do
+    let!(:board) {FactoryBot.create :board}
+    let!(:user) {FactoryBot.create :user}
+    let!(:user_board) {FactoryBot.create :user_board, board_id: board.id, user_id: user.id}
+    let!(:list) {FactoryBot.create :list, board_id: board.id}
+    let(:tag) {FactoryBot.create :tag, list_id: list.id}
+    let(:tag_last) {FactoryBot.create :tag, list_id: list.id}
+
+    context "opened" do
+      before {tag.closed = false}
+      it "is expected to return opened tags" do
+        expect(Tag.opened).to include tag
+      end
+    end
+
+    context "closed" do
+      before {tag.closed = false}
+      it "is expected to return closed tags" do
+        expect(Tag.opened).to include tag
+      end
+    end
+
+    context "last_position" do
+      before do
+        tag.position = 0
+        tag_last.position = 1
+      end
+      it "is expected to return the tag at the last position" do
+        expect(Tag.opened).to include tag_last
+      end
+    end
+
+    context "order_by_updated_at" do
+      before do
```

Comment:
before để thế này nghĩa là sao e

Reply:
em dùng thế này để thực hiện nhiều câu lệnh ạ, vì chỗ này em test sắp xếp theo updated_at nên phải cần ít nhất 2 bản ghi

Reply:
có những loại before nào e ?

Reply:
Có 2 loại before ạ:
:context (:all) - chạy 1 lần duy nhất trước khi thực hiện các test case
:example (:each) - chạy trước các test case, sẽ reset biến instance trong before mỗi khi khối it được thực thi